In [1]:
import pandas as pd
import numpy as np

In [ ]:
#download the class movie reviews from here:
#https://docs.google.com/spreadsheets/d/17rCJzmWxqvAu9rkpkgt4ToccIlY4A1Ffuu1W9X3B8Ag/
#then read that in as a pandas dataframe

In [2]:
M = pd.read_csv("Movie Reviews.csv", index_col='Name')

In [3]:
def pearson(s1, s2):
    """Take two pd.Series objects and return a pearson correlation."""
    s1_c = s1 - s1.mean()
    s2_c = s2 - s2.mean()
    return np.sum(s1_c * s2_c) / np.sqrt(np.sum(s1_c ** 2) * np.sum(s2_c ** 2))

In [4]:
def get_recs(movie_name, M, num):

    import numpy as np
    reviews = []
    for title in M.columns:
        if title == movie_name:
            continue
        cor = pearson(M[movie_name], M[title])
        if np.isnan(cor):
            continue
        else:
            reviews.append((title, cor))
    
    reviews.sort(key=lambda tup: tup[1], reverse=True)
    return reviews[:num]

    

###Question 1:  What movie is most Similar to 'The fault in our stars' (60 pts)


In [6]:
recs = get_recs('The Fault in Our Stars', M, 10)
recs[:10]

[('Malificent', 0.4290694760646056),
 ('Godzilla', 0.32655088533483884),
 ('How to Train your Dragon 2', 0.21246439225199001),
 ('Interstellar', 0.14553965767795929),
 ('Gone Girl', 0.10292874720773869),
 ('Divergent', 0.077957424633920691),
 ('Unbroken', 0.074198141493007602),
 ('Guardians of the Galaxy', 0.066871874009248947),
 ('Transformers', 0.064887439494061949),
 ('The Hunger Games: Mockingjay - Part 1', 0.063868558494407451)]

Question 1 Answer:  Malificent

###Question 2:  Which movie(s) would you most like to see, based on your classmates experience? (40 pts)

In [8]:
#HINTS...

#the movies I've seen
my_ratings = M.ix['David Reyling'].dropna()
my_ratings

The Hunger Games: Mockingjay - Part 1    1
The Hobbit                               1
Transformers                             1
Godzilla                                 3
Name: David Reyling, dtype: float64

In [9]:
#the movie I've rated the highest
my_ratings.sort(ascending=False)
my_ratings

C:\Users\David Reyling\Anaconda2\lib\site-packages\ipykernel\__main__.py:2: FutureWarning: sort is deprecated, use sort_values(inplace=True) for for INPLACE sorting
  from ipykernel import kernelapp as app


Godzilla                                 3
Transformers                             1
The Hobbit                               1
The Hunger Games: Mockingjay - Part 1    1
Name: David Reyling, dtype: float64

In [10]:
#the name of the movie I've rated the highest
fav_movie = my_ratings.index[0]
fav_movie

'Godzilla'

Question 2 Answer: Godzilla I rated Godzilla the highest. 

Just a side note though, I got my numbers backwards. I should have rate the 1's as 5's, and vice versa. Had I rated them properly I would have The Hobbit or Transformers as my favorite.


###Question 3: Bonus Question...  For all the movies you haven't seen, can you predict how you'd rate them using your the class reviews? (10 pts)

Question 3 Answer:  I will predict how I would rate Interstellar which I haven't seen.  I would do the exact same code for each particular movie I haven't seen.

In [15]:
recs = get_recs('Interstellar', M, 10)
recs

[('Unbroken', 0.30827889725078506),
 ('Malificent', 0.24099128506174256),
 ('Divergent', 0.24012818404319206),
 ('Transformers', 0.16711509642287781),
 ('Godzilla', 0.16468163539289821),
 ('The Fault in Our Stars', 0.14553965767795929),
 ('Gone Girl', 0.1031977194395927),
 ('300: Rise of an Empire', 0.079012375900903598),
 ('The Hobbit', 0.064027755269616335),
 ('How to Train your Dragon 2', 0.012246178545339669)]

I see Transformers has a correlation of 0.16712, and The Hobbit has a correlation of 0.06403.  This would tell me I probably would like the movie, but likely not as strong as Transformers or The Hobbit.  I am likely to give this particular movie a 3 or 4 rather than a 5.  Keep in mind, I should have rated Transformers and The Hobbit as a 5 instead of 1.